src/follow-focus.cpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright © 2020 Till Smejkal
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
+ * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <wayfire/plugin.hpp>
+#include <wayfire/core.hpp>
+#include <wayfire/option-wrapper.hpp>
+#include <wayfire/output.hpp>
+#include <wayfire/output-layout.hpp>
+#include <wayfire/view.hpp>
+#include <wayfire/util.hpp>
+
+
+class wayfire_follow_focus : public wf::plugin_interface_t
+{
+   private:
+    wf::wl_idle_call change_focus;
+
+    wf::option_wrapper_t<bool> should_change_view{"follow-focus/change_view"};
+    wf::option_wrapper_t<bool> should_change_output{"follow-focus/change_output"};
+
+    void change_view()
+    {
+        auto view = wf::get_core().get_cursor_focus_view();
+        wf::get_core().set_active_view(view);
```

Comment:
You should use `wf::get_core().focus_view()` instead of `set_active_view`. The former will also raise the view to the front and focus the view's output.

Reply:
No, I just wanted to change the keyboard focus without restacking the views. This is what many old X-based DE used to do.

Reply:
Then focus won't be restored to the last view, that's how things work. You could try to keep a list of focused views in the plugin if you want to ...

Reply:
Btw, are you using simple-tile for all views? I don't think not-raising makes much sense with floating views (why would I type in a window which is 90% covered by another ??)

Reply:
Hm, usually, if I want to raise the window to the front I can just simply click on it. I use the plugin's functionality really for this use case where I want keyboard focus on another view without raising it. For example when I read papers and take notes, I usually have this work flow. In this case the pdf-viewer has ca. 80% of my screen and in addition to that I have a terminal with vim open (in the background) where I take notes …
But I don't have very hard feelings about the actual behavior. If you think that raising the window makes more sense, I can change this.

Reply:
It might make sense to have two options in this case, raise-on-click and focus-follows-mouse, since raise and focus are separate.

Reply:
Raise-on-click is afaik already provided by the core, isn't it?

Reply:
The point would be to not raise if the option is disabled. Maybe it's not the greatest choice of option name.

Reply:
Just to make sure, I understand you correctly: You suggest to always raise the views on top when the pointer hovers over the window and add an option to change this behavior so that the window only gets keyboard focus but isn't raised on top?

Reply:
>You should use wf::get_core().focus_view() instead of set_active_view. The former will also raise the view to the front and focus the view's output.

>I just wanted to change the keyboard focus without restacking the views. This is what many old X-based DE used to do.

I am saying to make these two behaviors to be selected with a toggle option.

Reply:
I think it is not possible to raise a view (using focus_view()) without giving it keyboard focus. Or did I miss something? Hence, I would think only one option makes sense, namely this one:
> always raise the views on top when the pointer hovers over the window and add an option to change this behavior so that the window only gets keyboard focus but isn't raised on top?

Reply:
Yes, I think you're right.

Reply:
> I think it is not possible to raise a view (using focus_view()) without giving it keyboard focus. 

This is perfectly possible, look at `workspace-manager.hpp`. However, when something triggers refocusing, that front view will be refocused.

Reply:
After having a closer look at `workspace-manager.hpp`, I don't think that its functionality can help me. At the moment `focus_view` does what is necessary when I want the window to be raised and if not `set_active_view` does the correct thing as well. How to fix the refocusing issue and that fast-switcher works again as expected, has to be solved differently. Both assume that stacking order in the workspace is equal to the activation order. However, this is an assumption that I break. Maybe adding a note to the option is enough for now?